### ISSUE TITLE
Draw Png Image - TypeError: "this.value is undefined"

### ISSUE THREAD

**Issue by madmansn0w** [Hopding/pdf-lib#315]
```
TypeError: "this.value is undefined"
    sizeInBytes http://...module.js:53721
    sizeInBytes http://...module.js:53145
    computeObjectOffsets http://...module.js:56546
    PDFObjectStream http://...module.js:56495
    withContextAndObjects http://...module.js:56552
    computeBufferSize http://...module.js:57267
    step http://...vendor.js:195594
    verb http://...vendor.js:195575
    fulfilled http://...vendor.js:195565
    invoke http://...polyfills.js:6016
    onInvoke http://...vendor.js:99209
    invoke http://...polyfills.js:6015
    run http://...polyfills.js:5781
    scheduleResolveOrReject http://...polyfills.js:6512
    invokeTask http://...polyfills.js:6048
    onInvokeTask http://...vendor.js:99190
    invokeTask http://...polyfills.js:6047
    runTask http://...polyfills.js:5825
    drainMicroTaskQueue http://...polyfills.js:6216
    invokeTask http://...polyfills.js:6126
    invoke http://...polyfills.js:6111
    timer http://...polyfills.js:8307
```
Stack trace shows drawImage() fails at PDFString.ts method:
```typescript
sizeInBytes(): number {
    return this.value.length + 2;
  }
```
which was called from:
```typescript
PDFDict.prototype.sizeInBytes = function () {
        var size = 5;
        var entries = this.entries();
        for (var idx = 0, len = entries.length; idx < len; idx++) {
            var _a = entries[idx], key = _a[0], value = _a[1];
            size += key.sizeInBytes() + value.sizeInBytes() + 2;
        }
        return size;
    };
```
Going into the second method, I have these values:
len: 14
key: { encodedName: '/V' }
value : { value: undefined }
entries and _a are arrays.

I'm not certain what the issue is. I'm working out of Angular 8 / typescript. 

The relevant code:
```typescript
const sigContractBytes = data.signatureContract;
const sigContractImage = await pdfDoc.embedPng(sigContractBytes as Uint8Array);
const sigContractDims = sigContractImage.scale(0.75);
...
pages[9]
    .drawImage(sigContractImage, {
      x: 90,
      y: 160,
      width: sigContractDims.width,
      height: sigContractDims.height
    });
```

**Comment by madmansn0w** [comment]
Ok, in the course of trying to figure out what was happening, I defined all of my fields with values of 'undefined' and everything works as expected now. Not sure what to make of this, considering that I was trying to draw images and it turned out to be an undefined text field value, which is unrelated at first glance.

**Comment by Hopding** [comment]
@madmansn0w Can you provide some code and other assets so I can try to reproduce the issue? I'm afraid I can't really help you much with the information you've provided thus far.

**Comment by madmansn0w** [comment]
Sure, I can try to explain, however, I have a new issue (I will open a different thread for it shortly.)
I'll start with the code I have with this particular issue (which is now resolved, but was not easy to figure out.)

```typescript
async getPdf(data) {
    var pdfUrl = '@/../assets/_pdf/NewHires.pdf';
    const pdfBuffer = await fetch(pdfUrl).then((res) => res.arrayBuffer());
    const pdfDoc = await PDFDocument.load(pdfBuffer);
    pdfDoc.registerFontkit(fontkit);
    // const FontHelvetica = pdfDoc.embedFont('Helvetica');
    const getAcroForm = pdfDoc => {
      return pdfDoc.catalog.lookup(PDFName.of('AcroForm'));
    };

    const getAcroFields = (pdfDoc) => {
      if (!pdfDoc.catalog.get(PDFName.of('AcroForm'))) return [];
      const acroForm = pdfDoc.context.lookup(pdfDoc.catalog.get(PDFName.of('AcroForm')));
      if (!acroForm.get(PDFName.of('Fields'))) return [];
      const acroFields = acroForm.context.lookup(acroForm.get(PDFName.of('Fields')));
      return acroFields.array.map(ref => acroForm.context.lookup(ref));
    };
    const findAcroFieldByName = (pdfDoc, name) => {
      const acroFields = getAcroFields(pdfDoc);
      return acroFields.find((acroField) => {
        const fieldName = acroField.get(PDFName.of('T'));
        return !!fieldName && fieldName.value === name;
      });
    };
    const fillAcroTextField = (
      // pdfDoc,
      acroField,
      // fontObject,
      text,
      // fontSize = 10,
    ) => {
      // console.log(acroField + ': ', PDFString.of(text));
      acroField.set(PDFName.of('V'), PDFString.of(text));
      acroField.set(PDFName.of('Ff'), PDFNumber.of(
        1 << 0 // Read Only
        |
        1 << 12 // Multiline
      ));
    };
    const fillInField = (
      pdfDoc, 
      fieldName, 
      text, 
      // fontSize
      ) => {
      const field = findAcroFieldByName(pdfDoc, fieldName);
      if (!field) throw new Error(`Missing AcroField: ${fieldName}`);

      fillAcroTextField(field, text);
    };
    const logAcroFieldNames = (pdfDoc: PDFDocument) => {
      const acroFields = getAcroFields(pdfDoc);
      acroFields.forEach((acroField) => {
        console.log(
          'Field Name:',
          acroField.get(PDFName.of('T')).toString(),
          'Field Type:',
          acroField.get(PDFName.of('FT')).toString()
        );
      });
    };

    const acroForm = getAcroForm(pdfDoc);
    const acroFields = getAcroFields(pdfDoc);
    const fieldNames = {
      rate: 'rate',
      ( ... )
      jointName: 'jointName'
    };

    console.log('writing fields ...')
    fillInField(pdfDoc, fieldNames.rate, data.rate || '');
    (...)
    fillInField(pdfDoc, fieldNames.jointName, data.jointName);

    // Boolean field setter
    acroForm.set(PDFName.of('new'), data.action == 'new' ? PDFBool.True : PDFBool.False);

    const pages = pdfDoc.getPages();
    console.log('embedding ...');

    const sigContractBytes = data.signatureContract;
    const sigContractImage = await pdfDoc.embedPng(sigContractBytes as Uint8Array);
    const sigContractDims = sigContractImage.scale(0.75);

    console.log('drawing images ...');
    pages[9]
    .drawImage(sigContractImage, {
      x: 90,
      y: 160,
      width: sigContractDims.width,
      height: sigContractDims.height
    });
    const pdfBytes = await pdfDoc.save();
    // ArrayBuffer -> Blob
    // var uint8Array          = new Uint8Array(pdfBytes);
    var arrayBuffer             = pdfBytes.buffer;
    var blob                       = new Blob([arrayBuffer], { type: 'application/pdf' } );
    var hireDoc                  = this.blobToFile(blob, 'hireDoc.pdf');
    return hireDoc;
  }
  public blobToFile = (theBlob: Blob, fileName: string): File => {
    var b: any = theBlob;
    // const fileArray: FileList[] = [];
    //A Blob() is almost a File() - it's just missing the two properties below which we will add
    b.lastModified = new Date().getTime();
    b.name = fileName;
    b.webkitRelativePath = "";
    b.lastModifiedDate = new Date(b.lastModified);
    return new File([b], fileName, {type: 'application/pdf'});
  }
```

The above is (most) of the code, and for this particular error, it was related to 
```typescript
const fieldNames = {
      rate: 'rate',
      ( ... )
      jointName: 'jointName'
    };
```
having undefined values as inputted from a submitted data form. 
Because they were undefined, when I tried to execute 

```typescript
const sigContractBytes = data.signatureContract;
const sigContractImage = await pdfDoc.embedPng(sigContractBytes as Uint8Array);
const sigContractDims = sigContractImage.scale(0.75);
```
I would receive the TypeError "this.value is undefined."
At some point, in the process of handling images, there is a check to see if text fields have a defined value other than null or undefined.

What's weird about this is that it didn't catch the undefined text field values until attempting to embed the images. 

I hope that makes things clearer.

**Comment by Hopding** [comment]
@madmansn0w It sounds like you were able to resolve everything in https://github.com/Hopding/pdf-lib/issues/318? Is it okay to close this issue now?

**Comment by madmansn0w** [comment]
If you feel that what I've shared here doesn't point out to any bugs, and it is all working as expected, then sure. I just felt that there were some issues in getting it working, and I'm not entirely certain why. 

Thanks @Hopding

**Comment by jdc-cunningham** [comment]
Adding a comment, somewhat related. I had a problem where embedded images were not appearing on all pages of the doc. I was embedding them per page on the doc as needed. I switched it to pre-embedding all of the images to be used on the doc immediately before drawing onto the subsequent pages. That seems to have worked out much better.

To be clear: two images, one per page of same doc. I would embed image 1 on page 1,  image 2 on page 2. When the PDF was generated, the image on page 2 was missing... so now I just embed both images to the doc right away, before I iterate over the pages and apply.
